### PR TITLE
Karmada metrics adapter deployment script

### LINKED ISSUE #3625: feat: provide the script to deploy karmada-metrics-adapter
**What would you like to be added**:
* Provide a script to deploy `karmada-metrics-adapter`, similar to the one for `karmada-estimator`. You can find an example here: https://github.com/karmada-io/karmada/blob/master/hack/deploy-scheduler-estimator.sh
* Make sure to install `karmada-metrics-adapter` when running `hack/local-up-karmada.sh`.

**Why is this needed**:
Facilitate the deployment of `karmada-metrics-adapter` for users to utilize FederatedHPA and scale workloads across multiple clusters.

### PR DESCRIPTION
**Pull request by sbdtu5498**
**What type of PR is this?**
/kind feature

**What this PR does**:
Provides deployment scripts for karmada-metrics-adapter.

**Which issue(s) this PR fixes**:
Fixes #3625 

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

File: hack/deploy-metrics-adapter.sh [hunk 1]
```diff
@@ -0,0 +1,67 @@
+#!/bin/bash
+
+set -o errexit
+set -o nounset
+
+REPO_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
+source "${REPO_ROOT}"/hack/util.sh
+function usage() {
+  echo "This script will deploy karmada-metrics-adapter on host cluster"
+  echo "Usage: hack/deploy-metrics-adapter.sh  <HOST_CLUSTER_KUBECONFIG> <HOST_CLUSTER_NAME>>"
```
Comment (jwcesign) [inline hack/deploy-metrics-adapter.sh:10]: I think we need to specify the karmada-apiserver kubeconfig and context name

File: hack/deploy-metrics-adapter.sh [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+echo "using kubeconfig: "$KUBECONFIG
+
+# deploy karmada-metrics-adapter
+kubectl --context="${HOST_CLUSTER_NAME}" apply -f "${REPO_ROOT}/artifacts/deploy/karmada-metrics-adapter.yaml"
+
+# make sure that karmada-metrics-adapter is ready
+util::wait_pod_ready "${HOST_CLUSTER_NAME}" "${KARMADA_METRICS_ADAPTER_LABEL}" "${KARMADA_SYSTEM_NAMESPACE}"
+
+# deploy karmada-metrics-adapter-apiservice
+kubectl --context="karmada-apiservice" apply -f "${REPO_ROOT}/artifacts/deploy/karmada-metrics-adaptor-apiservice.yaml"
```
Comment (jwcesign) [inline hack/deploy-metrics-adapter.sh:51]: We are uncertain whether the context name should be `karmada-apiserver`. This parameter should be specified as an input flag.

**Review by jwcesign — COMMENTED** [review]
Thanks for your contribution, @sbdtu5498 . There are a few issues to address, but other than that, it looks good to me (LGTM).

**Comment by jwcesign** [comment]
By the way, please run `chmod 755 hack/xxx` to grant the necessary permissions for the script.

File: hack/local-up-karmada.sh [hunk 3]
```diff
@@ -126,13 +126,17 @@ kind load docker-image "${REGISTRY}/karmada-scheduler:${VERSION}" --name="${HOST
...
+kind load docker-image "${REGISTRY}/karmada-metrics-adapter:${VERSION}" --name="${HOST_CLUSTER_NAME}"
 kind load docker-image "${REGISTRY}/karmada-aggregated-apiserver:${VERSION}" --name="${HOST_CLUSTER_NAME}"
 kind load docker-image "${REGISTRY}/karmada-search:${VERSION}" --name="${HOST_CLUSTER_NAME}"
 
 #step5. install karmada control plane components
 "${REPO_ROOT}"/hack/deploy-karmada.sh "${MAIN_KUBECONFIG}" "${HOST_CLUSTER_NAME}"
 
-#step6. wait until the member cluster ready and join member clusters
+#step6. deploy karmada-metrics-adapter
+"${REPO_ROOT}"/hack/deploy-metrics-adapter.sh "${MAIN_KUBECONFIG}" "${HOST_CLUSTER_NAME}"
```
Comment (jwcesign) [inline hack/local-up-karmada.sh:137]: Do we need to install metrics-adapter by default when running `hack/local-up-karmada.sh`? cc @XiShanYongYe-Chang for checking

Comment (XiShanYongYe-Chang) [reply hack/local-up-karmada.sh:137]: Can we just provide the script and let the user decide whether to install it?

Comment (jwcesign) [reply hack/local-up-karmada.sh:137]: +1

Comment (sbdtu5498) [reply hack/local-up-karmada.sh:137]: Should we remove it then?

Comment (jwcesign) [reply hack/local-up-karmada.sh:137]: I prefer to delete it

File: hack/deploy-metrics-adapter.sh (hunk 1 above)
Comment (sbdtu5498) [reply hack/deploy-metrics-adapter.sh:10]: The karmada-apiserver context is causing too much confusion. I think the kubeconfig would be the same as the host-cluster-kubeconfig and the context karmada-apiserver is being used as sort of control parameter for permissions and security purpose.

File: hack/deploy-metrics-adapter.sh (hunk 2 above)
Comment (sbdtu5498) [reply hack/deploy-metrics-adapter.sh:51]: The logic behind taking karmada-apiserver as the context was that Karmada is using two different contexts i.e. karmada-apiserver and karmada-host that seem to be referring to the same entity. All the Kubectl request needs to be passed to the cluster where the Karmada control plane is hosted and aggregation should be with the API server of that cluster. In this case, to me, it seems both the karmada-apiserver and karmada-host are referring to the same entity. I took context as karmada-apiserver as it looked more of a convention to me as it was being used by both the API service for aggregation purposes. I wasn't sure either whether it is being used for security purposes or something else, so went with the convention.

Comment (sbdtu5498) [reply hack/deploy-metrics-adapter.sh:51]: There seems to be a small typo, "karmada-apiservice" -> "karmada-apiserver". I will fix it.

File: hack/local-up-karmada.sh (hunk 3 above)
Comment (sbdtu5498) [reply hack/local-up-karmada.sh:137]: Sure!

File: hack/deploy-metrics-adapter.sh (hunk 2 above)
Comment (jwcesign) [reply hack/deploy-metrics-adapter.sh:51]: Karmada is installed is a specific cluster(host cluster), it contains `kube-apiserver` component, named `karmada-apiserver`. So karmada-host is used to connect to the host cluster's kube-apiserver, karmada-apiserver is used to connect to `karmada-apiserver`. They are different entity.

**Comment by jwcesign** [comment]
Hello, @sbdtu5498! You can pass the DCO checking by running `git commit --amend -s`.

File: hack/deploy-metrics-adapter.sh (hunk 2 above)
Comment (sbdtu5498) [reply hack/deploy-metrics-adapter.sh:51]: I am still a bit confused, as it is a part of the same cluster kube-apiserver itself is capable of handling all such requests easily. Then why we are using two contexts here? I am not sure whether any permission boundaries exist or need to exist in the case of aggregated API.

File: hack/deploy-metrics-adapter.sh (hunk 1 above)
Comment (sbdtu5498) [reply hack/deploy-metrics-adapter.sh:10]: The point is, karmada-apiserver is just an extension of kube-apiserver and in that case, even if we had a single context only for the HOST_CLUSTER, wouldn't it still work in the same way?

Comment (jwcesign) [reply hack/deploy-metrics-adapter.sh:10]: The architecture looks like K8s on K8s:
image

karmada-apiserver context is used to connect to the kube-apiserver in karmada-system.
karmada-host is used to connect to the kube-apiserver in the host cluster.

**Comment by sbdtu5498** [comment]
> Hello, @sbdtu5498! You can pass the DCO checking by running `git commit --amend -s`.

Oh, shoot! did I miss it somewhere?

File: hack/deploy-metrics-adapter.sh (hunk 2 above)
Comment (jwcesign) [reply hack/deploy-metrics-adapter.sh]: Karmada is responsible for workload propagation, if we use the clusters' apiserver, the workload will be created in the cluster, but what we want is: the deployment should be created in member clusters. So we have to create ns, deploy another K8s there, and cut some functions for not creating real workloads there, just sync it to member clusters with policy.

File: hack/deploy-metrics-adapter.sh (hunk 1 above)
Comment (sbdtu5498) [reply hack/deploy-metrics-adapter.sh:10]: Ah, this makes more sense. It is namespaced in the same cluster for security and flexibility purposes.

**Comment by jwcesign** [comment]
Hi @sbdtu5498 
There are only tow nits:
1. fix DOC checking with `git commit --amend -s`
2. Define the flags to specify the kubeconfig file path and context name for karmada-apiserver

**Comment by sbdtu5498** [comment]
@jwcesign I just checked, DCO problem is persisting because a commit that came from reverting doesn't include the standard DCO comment that is probably being checked by the DCO bot, though it is signed. I will fix it. And regarding the kubeconfig file, I think we don't need a new kubeconfig file. It is totally possible for the same kubeconfig file to work differently with different contexts and I just checked hack/deploy_karmada.sh, they are also only passing HOSTC_CLUSTER_CONFIG as a parameter. So I don't think it should be a requirement. Please confirm.

**Comment by jwcesign** [comment]
> So I don't think it should be a requirement. Please confirm.

Hi @sbdtu5498 
I think it's ok if we only specify only one Kubeconfig.

**Comment by jwcesign** [comment]
By the way, you need to squash all the commits into one.

**Comment by RainbowMango** [comment]
> By the way, you need to squash all the commits into one.

Here is the guide for squash commits:
https://karmada.io/docs/contributor/github-workflow#squash-commits

**Review by jwcesign — COMMENTED** [review]
Great job! Thank you for your contribution! @sbdtu5498 

/lgtm

cc @RainbowMango  for final checking

**Comment by sbdtu5498** [comment]
> Great job! Thank you for your contribution! @sbdtu5498

Thanks a lot, @jwcesign!

**Review by RainbowMango — COMMENTED** [review]
/assign
I will add it to my queue and will try to review it tomorrow.

**Review by RainbowMango — COMMENTED** [review]
We need to grant permission to this script, like what we did for other scripts:
```
-rwxr-xr-x  1 root root  2361 Jan 10 11:45 deploy-agent-and-estimator.sh
-rwxr-xr-x  1 root root  3840 Jan 10 11:45 deploy-karmada-agent.sh
-rwxr-xr-x  1 root root  1991 Jan 10 11:45 deploy-karmada-opensearch.sh
-rwxr-xr-x  1 root root 16546 Jun  5 19:28 deploy-karmada.sh
-rw-r--r--  1 root root  1999 Jun 15 17:36 deploy-metrics-adapter.sh   // we need 755
```

You can do that by following commands:
1. # chmod 755 hack/deploy-metrics-adapter.sh
2. # git add hack/deploy-metrics-adapter.sh
3. # git commit --amend
4. # git push origin -f

**Comment by sbdtu5498** [comment]
@RainbowMango done.

File: hack/deploy-metrics-adapter.sh [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+echo "using kubeconfig: "$KUBECONFIG
+
+# deploy karmada-metrics-adapter
+kubectl --context="${HOST_CLUSTER_NAME}" apply -f "${REPO_ROOT}/artifacts/deploy/karmada-metrics-adapter.yaml"
+
+# make sure that karmada-metrics-adapter is ready
+util::wait_pod_ready "${HOST_CLUSTER_NAME}" "${KARMADA_METRICS_ADAPTER_LABEL}" "${KARMADA_SYSTEM_NAMESPACE}"
+
+# deploy karmada-metrics-adapter-apiservice
+kubectl --context="karmada-apiserver" apply -f "${REPO_ROOT}/artifacts/deploy/karmada-metrics-adaptor-apiservice.yaml"
```
Comment (jwcesign) [inline hack/deploy-metrics-adapter.sh:51]:
```diff
 # deploy karmada-metrics-adapter-apiservice
-kubectl --context="karmada-apiserver" apply -f "${REPO_ROOT}/artifacts/deploy/karmada-metrics-adaptor-apiservice.yaml"
+kubectl --context="karmada-apiserver" apply -f "${REPO_ROOT}/artifacts/deploy/karmada-metrics-adapter-apiservice.yaml"
```

Comment (RainbowMango) [reply hack/deploy-metrics-adapter.sh:51]: What's the difference?

Comment (RainbowMango) [reply hack/deploy-metrics-adapter.sh:51]: What do you mean `after this patch`?

Comment (jwcesign) [reply hack/deploy-metrics-adapter.sh:51]: adaptor -> adapter.
After making this change, it worked perfectly in my testing environment.

Comment (sbdtu5498) [reply hack/deploy-metrics-adapter.sh:51]: I will fix the typo.

**Comment by sbdtu5498** [comment]
@jwcesign it has been fixed. I think it'd be good to merge now. Kindly review. Thanks!

**Comment by jwcesign** [comment]
I have tested it and it works perfectly fine. cc @RainbowMango for approving it.

**Review by RainbowMango — COMMENTED** [review]
/lgtm

I already approved it, by the way.

**Pull request merged**